### DISCUSSION TITLE
S3; Easy way to convert S3 lifecycle configuration in xml to sdk compliant BucketLifecycleConfiguration struct.

### DISCUSSION THREAD

**Discussion by Church-** [awslabs/aws-sdk-rust#593] [General]
Officially per [S3 docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-configuration-examples.html) you can send a compliant XML file via a PUT operation to update a bucket's lifecycle config.

From reading over docs.rs and the source code for the lifecycle builder module there doesn't seem to be an easy way to pass in a compliant xml string and build a LifecycleConfiguration object from it.

I'm curious if anyone has built such a way or have an idea aside from implementing `impl From<String>` for it myself.

**Comment by jdisanti** [comment]
In `aws-sdk-s3`, if you make a `PutBucketLifecycleConfiguration` call, it constructs that XML for you as part of sending the operation. So it takes whichever lifecycle rules you add to it with the builder, and does that serialization for you. The SDK doesn't really have an easy way of overriding this so that you can replace the request body with your own XML payload.

What is your use-case for doing this in XML rather than constructing the rules via the SDK?

**Reply by Church-** [reply]
@jdisanti I'm writing a utility that generates an s3 lifecycle policiy via a retention policy for an already running daaemon.
I was thinking it'd have two subcommands, one to generate the compliant XML policy and write it to disk so a user could then upload that to AWS however they wished, and an optional upload command that would also handle uploading the policy file to AWS without the user needing to figure that out.

Worst comes to worst I can just write some structs and methods to reverse the XML policy file and generate the Lifecycle Rules the way the s3 sdk currently asks for. But of course if I didn't need to I'd rather not have to
